compiler/src/lnntoamm/Types.ts

```diff
@@ -197,83 +252,426 @@ class Struct extends Type {
-    return null;
+    const name = ast.get('variable').t.trim();
+    let work = ast.get('interfacedef');
+    if (work.has('interfacebody')) {
+      work = work.get('interfacebody').get('interfacelist');
+      const lines = [
+        work.get('interfaceline'),
+        ...work.get('cdr').getAll().map(cdr => cdr.get('interfaceline')),
+      ];
+      let fields: HasField[] = [];
+      let methods: HasMethod[] = [];
+      let operators: HasOperator[] = [];
+      lines.forEach(line => {
+        if (line.has('propertytypeline')) {
+          fields.push(HasField.fromPropertyTypeLine(line.get('propertytypeline'), scope));
+        } else if (line.has('functiontypeline')) {
+          methods.push(HasMethod.fromFunctionTypeLine(line.get('functiontypeline'), scope, name));
+        } else if (line.has('operatortypeline')) {
+          operators.push(HasOperator.fromOperatorTypeLine(line.get('operatortypeline'), scope, name));
+        } else {
+          throw new Error(`invalid ast: ${work}`);
+        }
+      });
+      return new Interface(
+        name,
+        ast,
+        fields,
+        methods,
+        operators,
+      );
+    } else if (work.has('interfacealias')) {
+      TODO('interface aliases')
+    } else {
+      throw new Error(`invalid ast: ${work}`);
+    }
   }
 
   breakdown(): Builtin {
+    if (this.tempDelegate !== null) {
+      return this.tempDelegate.breakdown();
+    }
     throw new Error(`interfaces cannot be broken down, and no concrete type was specified`);
   }
 
-  compatibleWithConstraint(ty: Type): boolean {
-    return TODO();
+  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
+    if (this.tempDelegate !== null) {
+      return this.tempDelegate.compatibleWithConstraint(ty, scope);
+    }
+    if (ty instanceof Builtin || ty instanceof Struct) {
+      if (ty instanceof Builtin && this.fields.length !== 0) {
+        // if ty is a Builtin and there are field requirements,
+        // then this interface doesn't apply.
+        // TODO: this might change depending on the opaque types
+        // we introduce
+        return false;
+      } else if (ty instanceof Struct) {
+        // ty is a Struct, ensure it has the same fields
+        // if (!this.fields.every(field => ty.fields[field.name] && ty.fields[field.name].eq(field.ty))) {
+        //   return false;
+        // }
+        if (!this.fields.every(field => Has.field(field, ty))) {
+          return false;
+        }
+      }
+
+      // check methods
+      return this.methods.every(m => Has.method(m, scope, ty)) &&
+        this.operators.every(o => Has.operator(o, scope, ty));
+      // check operators
+    } else if (ty instanceof Interface) {
+      // ensure `ty ⊆ this`
+      return ty.fields.every(field => this.fields.find(f => f.eq(field))) &&
+        ty.methods.every(method => this.methods.find(m => m.eq(method))) &&
+        ty.operators.every(operator => this.operators.find(o => o.eq(operator)));
```

Comment:
Nonblocking: It may make sense to change these from arrays to hashmaps for better lookup performance (this is currently `O(n*m)`) Since you can get an iterable array of the values with [`Object.values()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values) cheaply, it doesn't seem like it would impact any of the other uses, either?

Reply:
This is a relatively large refactor on the `Types.ts` file, so I'll leave it as a TODO and get to it later when optimizing the compiler